File: client/src/components/UserPanel.js
import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { jwtDecode } from 'jwt-decode';
import '../styles/UserPanel.css';

function UserPanel() {
    const navigate = useNavigate();
    const [activeTab, setActiveTab] = useState('profile');
    const [profile, setProfile] = useState(null);
    const [stats, setStats] = useState(null);
    const [favorites, setFavorites] = useState([]);
    const [email, setEmail] = useState('');
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword, setNewPassword] = useState('');
    const [confirmPassword, setConfirmPassword] = useState('');
    const [loading, setLoading] = useState(false);
    const token = localStorage.getItem('token');

    useEffect(() => {
        if (!token) {
            navigate('/login');
            return;
        }

        try {
            const decodedToken = jwtDecode(token);
            const currentTime = Date.now() / 1000;

            if (decodedToken.exp < currentTime) {
                localStorage.clear();
                showNotification('Sesja wygasła. Zaloguj się ponownie.', 'error');
                window.location.href = '/login';
                return;
            }

            if (!decodedToken.user_id) {
                navigate('/login');
                return;
            }

            // Sprawdź czy to nie jest pracownik
            if (decodedToken.rola === 'Bibliotekarz') {
                navigate('/worker-panel');
                return;
            }

            // Pobierz dane w zależności od aktywnej zakładki
            fetchData();
        } catch (err) {
            console.error('Invalid token:', err);
            showNotification('Nieprawidłowy token. Zaloguj się ponownie.', 'error');
            navigate('/login');
        }
    }, [navigate, activeTab, token]);

    const fetchData = async () => {
        setLoading(true);
        try {
            if (activeTab === 'profile') {
                await fetchProfile();
                await fetchStats();
            } else if (activeTab === 'favorites') {
                await fetchFavorites();
            }
        } catch (error) {
            console.error('Error fetching data:', error);
            showNotification('Błąd podczas pobierania danych', 'error');
        } finally {
            setLoading(false);
        }
    };
    
    const handleBookClick = (book) => {
      if (book.slug) {
        navigate(`/book/${book.slug}`);
      } else {
        console.error('No slug available for book:', book.title);
      }
    };

    const fetchProfile = async () => {
        try {
            const response = await fetch('http://localhost:3000/user/profile', {
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            const data = await response.json();
            if (response.ok) {
                setProfile(data);
            } else {
                console.error('Profile fetch error:', data.error);
                showNotification('Błąd podczas pobierania profilu', 'error');
            }
        } catch (error) {
            console.error('Profile fetch error:', error);
            showNotification('Błąd podczas pobierania profilu', 'error');
        }
    };

    const fetchStats = async () => {
        try {
            const response = await fetch('http://localhost:3000/user/stats', {
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            const data = await response.json();
            if (response.ok) {
                setStats(data);
            } else {
                console.error('Stats fetch error:', data.error);
            }
        } catch (error) {
            console.error('Stats fetch error:', error);
        }
    };

    const fetchFavorites = async () => {
        try {
            const response = await fetch('http://localhost:3000/user/favorites', {
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            const data = await response.json();
            if (response.ok) {
                setFavorites(data.favorites);
            } else {
                console.error('Favorites fetch error:', data.error);
                showNotification('Błąd podczas pobierania ulubionych', 'error');
            }
        } catch (error) {
            console.error('Favorites fetch error:', error);
            showNotification('Błąd podczas pobierania ulubionych', 'error');
        }
    };

    const showNotification = (message, type = 'success') => {
        const notification = document.createElement('div');
        notification.className = `user-notification user-notification-${type}`;
        notification.innerHTML = `
            <div class="user-notification-content">
                <span class="user-notification-icon">${type === 'success' ? '✓' : '⚠'}</span>
                <span class="user-notification-message">${message}</span>
            </div>
        `;
        
        if (!document.querySelector('#user-notification-styles')) {
            const style = document.createElement('style');
            style.id = 'user-notification-styles';
            style.textContent = `
                .user-notification {
                    position: fixed;
                    top: 20px;
                    right: 20px;
                    z-index: 10000;
                    min-width: 300px;
                    padding: 1rem 1.5rem;
                    border-radius: 12px;
                    backdrop-filter: blur(10px);
                    box-shadow: 0 8px 32px rgba(0, 0, 0, 0.3);
                    animation: userSlideIn 0.3s ease-out, userFadeOut 0.3s ease-in 2.7s;
                    animation-fill-mode: forwards;
                }
                .user-notification-success {
                    background: linear-gradient(135deg, rgba(46, 160, 67, 0.9), rgba(40, 167, 69, 0.9));
                    border: 1px solid rgba(46, 160, 67, 0.3);
                }
                .user-notification-error {
                    background: linear-gradient(135deg, rgba(220, 53, 69, 0.9), rgba(176, 42, 55, 0.9));
                    border: 1px solid rgba(220, 53, 69, 0.3);
                }
                .user-notification-content {
                    display: flex;
                    align-items: center;
                    gap: 0.75rem;
                    color: white;
                    font-weight: 500;
                }
                .user-notification-icon {
                    font-size: 1.2rem;
                    font-weight: bold;
                }
                @keyframes userSlideIn {
                    from { transform: translateX(100%); opacity: 0; }
                    to { transform: translateX(0); opacity: 1; }
                }
                @keyframes userFadeOut {
                    from { opacity: 1; }
                    to { opacity: 0; }
                }
            `;
            document.head.appendChild(style);
        }
        
        document.body.appendChild(notification);
        
        setTimeout(() => {
            if (notification.parentNode) {
                notification.parentNode.removeChild(notification);
            }
        }, 3000);
    };

    const handleLogout = () => {
        if (window.confirm('Czy na pewno chcesz się wylogować?')) {
            localStorage.clear();
            showNotification('Zostałeś wylogowany', 'success');
            setTimeout(() => {
                window.location.href = '/login';
            }, 1000);
        }
    };

    const updateEmail = async () => {
        if (!email.trim()) {
            showNotification('Wprowadź nowy email', 'error');
            return;
        }

        try {
            const response = await fetch('http://localhost:3000/user/email', {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${token}`,
                },
                body: JSON.stringify({ email }),
            });
            const data = await response.json();
            
            if (data.error) {
                showNotification(data.error, 'error');
            } else {
                showNotification(data.message || 'Email został zaktualizowany', 'success');
                setEmail('');
                fetchProfile();
            }
        } catch (error) {
            showNotification('Błąd podczas aktualizacji email', 'error');
        }
    };

    const updatePassword = async () => {
        if (!oldPassword.trim() || !newPassword.trim()) {
            showNotification('Wypełnij wszystkie pola hasła', 'error');
            return;
        }

        if (newPassword !== confirmPassword) {
            showNotification('Nowe hasła się nie zgadzają', 'error');
            return;
        }

        if (newPassword.length < 6) {
            showNotification('Nowe hasło musi mieć co najmniej 6 znaków', 'error');
            return;
        }

        try {
            const response = await fetch('http://localhost:3000/user/password', {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${token}`,
                },
                body: JSON.stringify({
                    old_password: oldPassword,
                    new_password: newPassword,
                }),
            });
            const data = await response.json();
            
            if (data.error) {
                showNotification(data.error, 'error');
            } else {
                showNotification(data.message || 'Hasło zostało zmienione', 'success');
                setOldPassword('');
                setNewPassword('');
                setConfirmPassword('');
            }
        } catch (error) {
            showNotification('Błąd podczas zmiany hasła', 'error');
        }
    };

    const removeFromFavorites = async (bookId) => {
        if (!window.confirm('Czy na pewno chcesz usunąć tę książkę z ulubionych?')) {
            return;
        }

        try {
            const response = await fetch(`http://localhost:3000/user/favorites/${bookId}`, {
                method: 'DELETE',
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            const data = await response.json();
            
            if (response.ok) {
                showNotification(data.message || 'Książka została usunięta z ulubionych', 'success');
                fetchFavorites();
            } else {
                showNotification(data.error || 'Błąd podczas usuwania', 'error');
            }
        } catch (error) {
            showNotification('Błąd podczas usuwania książki z ulubionych', 'error');
        }
    };

    const renderLoadingSpinner = () => (
        <div className="user-loading-container">
            <div className="user-spinner"></div>
            <p>Ładowanie danych...</p>
        </div>
    );

    return (
        <div className="user-panel">
            <div className="user-sidebar">
                <h2>Panel Użytkownika</h2>
                <button 
                    className={activeTab === 'profile' ? 'active' : ''}
                    onClick={() => setActiveTab('profile')}
                >
                    👤 Profil
                </button>
                <button 
                    className={activeTab === 'edit' ? 'active' : ''}
                    onClick={() => setActiveTab('edit')}
                >
                    ⚙️ Edycja danych
                </button>
                <button 
                    className={activeTab === 'favorites' ? 'active' : ''}
                    onClick={() => setActiveTab('favorites')}
                >
                    ❤️ Ulubione książki
                </button>
                <button onClick={handleLogout} className="user-logout-btn">
                    🚪 Wyloguj
                </button>
            </div>
            
            <div className="user-content">
                {loading && renderLoadingSpinner()}
                
                {!loading && activeTab === 'profile' && profile && (
                    <div className="user-profile-section">
                        <div className="user-profile-header">
                            <div className="user-avatar">
                                {profile.name?.[0]?.toUpperCase() || 'U'}
                            </div>
                            <div className="user-profile-info">
                                <p><strong>ID:</strong> {profile.user_id}</p>
                                <p><strong>Imię:</strong> {profile.name}</p>
                                <p><strong>Email:</strong> {profile.email}</p>
                                <p><strong>Data rejestracji:</strong> {
                                    profile.created_at ? 
                                    new Date(profile.created_at).toLocaleDateString('pl-PL') : 
                                    'Brak danych'
                                }</p>
                            </div>
                        </div>
                        
                        {stats && (
                            <div className="user-stats-section">
                                <h3>📊 Twoje statystyki</h3>
                                <div className="user-stats-grid">
                                    <div className="user-stat-item">
                                        <span className="user-stat-number">{stats.favorites_count}</span>
                                        <span className="user-stat-label">Ulubione książki</span>
                                    </div>
                                    <div className="user-stat-item">
                                        <span className="user-stat-number">{stats.reviews_count}</span>
                                        <span className="user-stat-label">Napisane recenzje</span>
                                    </div>
                                    <div className="user-stat-item">
                                        <span className="user-stat-number">{stats.average_rating ? stats.average_rating.toFixed(1) : '0.0'}</span>
                                        <span className="user-stat-label">Średnia ocena</span>
                                    </div>
                                    <div className="user-stat-item">
                                        <span className="user-stat-number">{stats.recommendations_sent || 0}</span>
                                        <span className="user-stat-label">Wysłane rekomendacje</span>
                                    </div>
                                </div>
                            </div>
                        )}
                    </div>
                )}

                {!loading && activeTab === 'edit' && (
                    <div className="user-edit-section">
                        <h3>📧 Zmień email</h3>
                        <input
                            type="email"
                            placeholder="Wprowadź nowy email"
                            value={email}
                            onChange={(e) => setEmail(e.target.value)}
                        />
                        <button onClick={updateEmail}>Zaktualizuj email</button>

                        <h3>🔒 Zmień hasło</h3>
                        <input
                            type="password"
                            placeholder="Obecne hasło"
                            value={oldPassword}
                            onChange={(e) => setOldPassword(e.target.value)}
                        />
                        <input
                            type="password"
                            placeholder="Nowe hasło (min. 6 znaków)"
                            value={newPassword}
                            onChange={(e) => setNewPassword(e.target.value)}
                        />
                        <input
                            type="password"
                            placeholder="Powtórz nowe hasło"
                            value={confirmPassword}
                            onChange={(e) => setConfirmPassword(e.target.value)}
                        />
                        <button onClick={updatePassword}>Zmień hasło</button>
                    </div>
                )}

                {!loading && activeTab === 'favorites' && (
                    <div className="user-favorites-section">
                        <h3>📚 Twoje ulubione książki ({favorites.length})</h3>
                        {favorites.length === 0 ? (
                            <div className="user-no-favorites">
                                <p>📖 Nie masz jeszcze żadnych ulubionych książek.</p>
                                <p>Dodaj książki do ulubionych, aby je tutaj zobaczyć!</p>
                            </div>
                        ) : (
                            <div className="user-favorites-grid">
                                {favorites.map((book) => (
                                    <div key={book.book_id} className="user-book-card">
                                        <div className="user-book-cover">
                                            {book.cover_thumb ? (
                                                <img 
                                                    src={book.cover_thumb} 
                                                    alt={book.title}
                                                    onError={(e) => {
                                                        e.target.style.display = 'none';
                                                        e.target.parentNode.innerHTML = '<div class="user-no-cover">📖</div>';
                                                    }}
                                                />
                                            ) : (
                                                <div className="user-no-cover">📖</div>
                                            )}
                                        </div>
                                        <div className="user-book-info">
                                            <h4 className="user-book-title">{book.title}</h4>
                                            <p className="user-book-author">{book.author || 'Autor nieznany'}</p>
                                            {book.genres && book.genres.length > 0 && (
                                                <div className="user-book-genres">
                                                    {book.genres.map((genre, index) => (
                                                        <span key={index} className="user-genre-tag">
                                                            {genre}
                                                        </span>
                                                    ))}
                                                </div>
                                            )}
                                            <p className="user-added-date">
                                                Dodano: {new Date(book.added_at).toLocaleDateString('pl-PL')}
                                            </p>
                                            {book.api_error && (
                                                <p className="user-api-error">
                                                    ⚠️ Nie udało się pobrać pełnych danych
                                                </p>
                                            )}
                                        </div>
                                        <div className="user-book-actions">
                                            <button 
                                                onClick={() => handleBookClick(book)}
                                                className="user-read-button"
                                            >
                                                📖 Szczegóły
                                            </button>
                                            <button 
                                                onClick={() => removeFromFavorites(book.book_id)}
                                                className="user-remove-button"
                                            >
                                                🗑️ Usuń
                                            </button>
                                        </div>
                                    </div>
                                ))}
                            </div>
                        )}
                    </div>
                )}
            </div>
        </div>
    );
}

export default UserPanel;

File: flask-server/blueprints/userpanel.py
# flask-server/blueprints/userpanel.py
from flask import Blueprint, request, jsonify
import psycopg, jwt, datetime, requests
from functools import wraps
from config import Config

userpanel = Blueprint('userpanel', __name__)

# JWT Secret Key
SECRET_KEY = Config.SECRET_KEY

# Middleware: JWT Auth decorator
def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.headers.get('Authorization')
        if not token:
            return jsonify({"error": "Brak tokenu"}), 401
        try:
            if token.startswith("Bearer "):
                token = token.split(" ")[1]
            decoded = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])
            request.user_id = decoded.get("user_id")
            # Sprawdź czy to jest zwykły użytkownik (nie pracownik)
            if decoded.get("rola") == "Bibliotekarz":
                return jsonify({"error": "Brak uprawnień"}), 403
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token wygasł"}), 401
        except jwt.InvalidTokenError:
            return jsonify({"error": "Nieprawidłowy token"}), 401
        return f(*args, **kwargs)
    return decorated


# 1. Informacje o profilu użytkownika
@userpanel.route('/user/profile', methods=['GET'])
@token_required
def get_profile():
    user_id = request.user_id
    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                cur.execute("""
                    SELECT user_id, name, email, created_at 
                    FROM users 
                    WHERE user_id = %s
                """, (user_id,))
                data = cur.fetchone()
                if not data:
                    return jsonify({"error": "Użytkownik nie znaleziony"}), 404

                return jsonify({
                    "user_id": data[0],
                    "name": data[1],
                    "email": data[2],
                    "created_at": data[3].isoformat() if data[3] else None,
                    "profile_picture": None  # Placeholder
                }), 200
    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500


# 2a. Zmiana emaila
@userpanel.route('/user/email', methods=['PUT'])
@token_required
def update_email():
    user_id = request.user_id
    data = request.get_json()
    new_email = data.get("email")
    
    if not new_email:
        return jsonify({"error": "Nowy email jest wymagany"}), 400

    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                # Sprawdź czy email już istnieje
                cur.execute("""
                    SELECT user_id FROM users WHERE email = %s AND user_id != %s
                """, (new_email, user_id))
                if cur.fetchone():
                    return jsonify({"error": "Email już istnieje w systemie"}), 400

                # Zaktualizuj email
                cur.execute("""
                    UPDATE users
                    SET email = %s
                    WHERE user_id = %s
                """, (new_email, user_id))
                conn.commit()
                return jsonify({"message": "Email zaktualizowany pomyślnie"}), 200
    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500


# 2b. Zmiana hasła
@userpanel.route('/user/password', methods=['PUT'])
@token_required
def update_password():
    user_id = request.user_id
    data = request.get_json()
    old_password = data.get("old_password")
    new_password = data.get("new_password")

    if not old_password or not new_password:
        return jsonify({"error": "Wszystkie pola są wymagane"}), 400

    if len(new_password) < 6:
        return jsonify({"error": "Nowe hasło musi mieć co najmniej 6 znaków"}), 400

    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                # Sprawdź stare hasło (zakładając że używasz bcrypt/crypt)
                cur.execute("""
                    SELECT user_id FROM users
                    WHERE user_id = %s AND password = crypt(%s, password)
                """, (user_id, old_password))
                if not cur.fetchone():
                    return jsonify({"error": "Obecne hasło jest nieprawidłowe"}), 401

                # Zmień hasło
                cur.execute("""
                    UPDATE users
                    SET password = crypt(%s, gen_salt('bf'))
                    WHERE user_id = %s
                """, (new_password, user_id))
                conn.commit()

                return jsonify({"message": "Hasło zostało zmienione"}), 200

    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500


# 3. Pobierz ulubione książki użytkownika
@userpanel.route('/user/favorites', methods=['GET'])
@token_required
def get_favorite_books():
    user_id = request.user_id
    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                cur.execute("""
                    SELECT 
                        b.book_id,
                        b.slug,
                        f.added_at
                    FROM favorites f
                    JOIN "Books" b ON f.book_id = b.book_id
                    WHERE f.user_id = %s
                    ORDER BY f.added_at DESC
                """, (user_id,))
                
                favorites = cur.fetchall()
                
                if not favorites:
                    return jsonify({"favorites": []}), 200

                # Dla każdej książki pobierz szczegóły z API
                favorite_books = []
                for book in favorites:
                    book_id, slug, added_at = book
                    
                    # Pobierz szczegóły książki z zewnętrznego API
                    try:
                        # Przykład wywołania API - dostosuj URL do swojego API
                        api_url = f"https://wolnelektury.pl/api/books/{slug}/"
                        response = requests.get(api_url, timeout=5)
                        
                        if response.status_code == 200:
                            book_data = response.json()
                           
                            genres = [genre['name'] for genre in book_data.get("genres", [])]

                            # Wyciągnij potrzebne informacje
                            book_info = {
                                "book_id": book_id,
                                "slug": slug,
                                "title": book_data.get("title", "Nieznany tytuł"),
                                "author": book_data.get("author", "Nieznany autor"),
                                "cover": book_data.get("cover", None),
                                "cover_thumb": book_data.get("cover_thumb", None),
                                "simple_thumb": book_data.get("simple_thumb", None),
                                "genres": genres,
                                "kinds": book_data.get("kinds", []),
                                "epochs": book_data.get("epochs", []),
                                "url": book_data.get("url", ""),
                                "added_at": added_at.isoformat() if added_at else None
                            }
                            favorite_books.append(book_info)
                        else:
                            # Jeśli API nie odpowiada, dodaj podstawowe info
                            book_info = {
                                "book_id": book_id,
                                "slug": slug,
                                "title": f"Książka {slug}",
                                "author": "Nieznany autor",
                                "cover": None,
                                "cover_thumb": None,
                                "simple_thumb": None,
                                "genres": [],
                                "kinds": [],
                                "epochs": [],
                                "url": "",
                                "added_at": added_at.isoformat() if added_at else None,
                                "api_error": True
                            }
                            favorite_books.append(book_info)
                            
                    except requests.RequestException as e:
                        # W przypadku błędu API, dodaj podstawowe informacje
                        book_info = {
                            "book_id": book_id,
                            "slug": slug,
                            "title": f"Książka {slug}",
                            "author": "Nieznany autor",
                            "cover": None,
                            "cover_thumb": None,
                            "simple_thumb": None,
                            "genres": [],
                            "kinds": [],
                            "epochs": [],
                            "url": "",
                            "added_at": added_at.isoformat() if added_at else None,
                            "api_error": True
                        }
                        favorite_books.append(book_info)

                return jsonify({"favorites": favorite_books}), 200

    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500


# 4. Dodaj książkę do ulubionych
@userpanel.route('/user/favorites', methods=['POST'])
@token_required
def add_to_favorites():
    user_id = request.user_id
    data = request.get_json()
    book_id = data.get("book_id")
    
    if not book_id:
        return jsonify({"error": "ID książki jest wymagane"}), 400

    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                # Sprawdź czy książka już jest w ulubionych
                cur.execute("""
                    SELECT entry_id FROM favorites 
                    WHERE user_id = %s AND book_id = %s
                """, (user_id, book_id))
                
                if cur.fetchone():
                    return jsonify({"error": "Książka już jest w ulubionych"}), 400

                # Sprawdź czy książka istnieje
                cur.execute("""
                    SELECT book_id FROM "Books" WHERE book_id = %s
                """, (book_id,))
                
                if not cur.fetchone():
                    return jsonify({"error": "Książka nie istnieje"}), 404

                # Dodaj do ulubionych
                cur.execute("""
                    INSERT INTO favorites (user_id, book_id, added_at)
                    VALUES (%s, %s, CURRENT_DATE)
                """, (user_id, book_id))
                conn.commit()

                return jsonify({"message": "Książka dodana do ulubionych"}), 201

    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500


# 5. Usuń książkę z ulubionych
@userpanel.route('/user/favorites/<int:book_id>', methods=['DELETE'])
@token_required
def remove_from_favorites(book_id):
    user_id = request.user_id
    
    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                # Sprawdź czy książka jest w ulubionych
                cur.execute("""
                    SELECT entry_id FROM favorites 
                    WHERE user_id = %s AND book_id = %s
                """, (user_id, book_id))
                
                if not cur.fetchone():
                    return jsonify({"error": "Książka nie jest w ulubionych"}), 404

                # Usuń z ulubionych
                cur.execute("""
                    DELETE FROM favorites 
                    WHERE user_id = %s AND book_id = %s
                """, (user_id, book_id))
                conn.commit()

                return jsonify({"message": "Książka usunięta z ulubionych"}), 200

    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500


# 6. Pobierz statystyki użytkownika
@userpanel.route('/user/stats', methods=['GET'])
@token_required
def get_user_stats():
    user_id = request.user_id
    
    try:
        with psycopg.connect(**Config.get_db_connection_params()) as conn:
            with conn.cursor() as cur:
                # Liczba ulubionych książek
                cur.execute("""
                    SELECT COUNT(*) FROM favorites WHERE user_id = %s
                """, (user_id,))
                favorites_count = cur.fetchone()[0]

                # Liczba recenzji
                cur.execute("""
                    SELECT COUNT(*) FROM book_reviews WHERE user_id = %s
                """, (user_id,))
                reviews_count = cur.fetchone()[0]

                # Średnia ocena wystawiona przez użytkownika
                cur.execute("""
                    SELECT AVG(rating) FROM book_reviews WHERE user_id = %s
                """, (user_id,))
                avg_rating = cur.fetchone()[0]

                # Liczba rekomendacji wysłanych
                cur.execute("""
                    SELECT COUNT(*) FROM user_recommendations WHERE rec_by_user_id = %s
                """, (user_id,))
                recommendations_sent = cur.fetchone()[0]

                # Liczba rekomendacji otrzymanych
                cur.execute("""
                    SELECT COUNT(*) FROM user_recommendations WHERE user_id = %s
                """, (user_id,))
                recommendations_received = cur.fetchone()[0]

                return jsonify({
                    "favorites_count": favorites_count,
                    "reviews_count": reviews_count,
                    "average_rating": float(avg_rating) if avg_rating else 0,
                    "recommendations_sent": recommendations_sent,
                    "recommendations_received": recommendations_received
                }), 200

    except psycopg.Error as e:
        return jsonify({"error": str(e)}), 500